target_decisioning_engine/geo_provider.py

```diff
@@ -0,0 +1,202 @@
+    :return: (delivery_api_client.Model.geo.Geo) geo object
+    """
+    if not initial:
+        initial = Geo()
+
+    def geo_mapping_accumulator(result, geo_mapping):
+        """
+        :param result: (delivery_api_client.Model.geo.Geo) geo object
+        :param geo_mapping: (dict) mapping object for handling specific geo fields
+        :return: (delivery_api_client.Model.geo.Geo) updated result
+        """
+        value = value_fn(geo_mapping.get("header_name"))
+        if value:
+            value = geo_mapping.get("parse_value")(value) if geo_mapping.get("parse_value") else value
+            setattr(result, geo_mapping.get("value_key"), value)
+        return result
+
+    return reduce(geo_mapping_accumulator, GEO_MAPPINGS, initial)
+
+
+def create_or_update_geo_object(geo_data=None, existing_geo_context=None):
+    """
+    :param existing_geo_context: (delivery_api_client.Model.geo.Geo) geo object
+    :param geo_data: (dict) geo payload or headers to merge into existing_geo_context
+    :return: (delivery_api_client.Model.geo.Geo) geo object
+    """
+    if not geo_data:
+        geo_data = {}
+
+    def _get_payload(key):
+        return geo_data.get(key)
+
+    return _map_geo_values(_get_payload, existing_geo_context)
+
+
+def is_missing_geo_fields(geo_request_context):
+    """
+    :param geo_request_context: (delivery_api_client.Model.geo.Geo) geo object
+    :return: (bool) Returns True if specified geo fields are undefined, else False
+    """
+    return not geo_request_context.latitude and not geo_request_context.longitude and not \
+        geo_request_context.country_code and not geo_request_context.state_code and not geo_request_context.city
+
+
+class GeoProvider:
+    """GeoProvider"""
+
+    def __init__(self, config, artifact):
+        """
+        :param config: (target_decisioning_engine.types.decisioning_config.DecisioningConfig) config
+        :param artifact: (target_decisioning_engine.types.decisioning_artifact.DecisioningArtifact) artifact
+        """
+        self.logger = get_logger()
+        self.pool_manager = urllib3.PoolManager()
+        self.config = config
+        self.artifact = artifact
+        self.geo_targeting_enabled = artifact.get("geoTargetingEnabled", False)
+        self.event_emitter = config.event_emitter or noop
+
+    def _execute_request(self, geo_lookup_path, headers, queue):
+        """Sends http request for geo data"""
+        response = self.pool_manager.request(HTTP_GET, geo_lookup_path, headers=headers)
+        queue.put(response)
+
+    def _request_geo(self, geo_lookup_path, headers):
+        """Executes geo request in a new thread"""
+        result = Queue()
+        request_thread = Thread(target=self._execute_request,
+                                name="GeoProvider.request",
+                                args=[geo_lookup_path, headers, result])
+        request_thread.start()
+        request_thread.join()
```

Comment:
Since we have a join here, the calling thread is still blocking. Let's leave it is as for now, but add a separate task for future optimization to the backlog, so that we have end-to-end async flows (similar to NodeJS SDK implementation) and there are no blocking calls.